### ISSUE TITLE
Is it compatiable with kernel 3.2.16?

### ISSUE THREAD

**Issue by lxyscls** [iputils/iputils#274]
Hello, I want to compile ping with kernel header 3.2.16. But I found that there are too many redefines in linux/in6.h. It is so odd!

**Comment by kerolasa** [comment]
Could you attach linux/in6.h from your system so that I can have a look of it.

Also, Linux 3.2.16 is about eight years old. Keeping a project in state where it compiles without issues a little bit of historical burden is fine, but there is a limit how far in past support should reasonably reach out.

**Comment by pevik** [comment]
Either try to old version or try to define, what is missing yourself and send a PR. If it's reasonable we could use it.

**Comment by lxyscls** [comment]
@kerolasa 
Our environment is an embedded system, so it is hard to upgrade kernel for many other dependencies. But it is odd that I can build it with 3.10.x, an even older kernel on CentOS 7.6.

Here is the in6.h from 3.2.16, which I have change it to in6.txt for the upload limitation of Github.
[in6.txt](https://github.com/iputils/iputils/files/4777735/in6.txt)

Thank you very much!

**Comment by pevik** [comment]
Then post the build error. It might help just to add some definitions.

**Comment by lxyscls** [comment]
@pevik It is a little verbose.
```
/fortidev/bin/x86_64-linux-gnu-gcc -g -Wall -include config.h -include git-version.h -m64 -fno-omit-frame-pointer -Werror -Wno-address -fms-extensions -fno-strict-aliasing -fstack-protector-strong -fno-reorder-functions -fno-unit-at-a-time -Wno-unused-but-set-variable -fsigned-char -pipe -O2 -Wtrampolines  -Wno-pointer-sign -fgnu89-inline -ggdb -DCONFIG_SUITEB192 -MD -MP  -I/home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils -I/home/sks/lxy/fos2_2020-06-04_11_30 -I/home/sks/lxy/fos2_2020-06-04_11_30/cmf/include -I/home/sks/lxy/fos2_2020-06-04_11_30/migbase/include -I/home/sks/lxy/fos2_2020-06-04_11_30/fgtcommon -I/home/sks/lxy/fos2_2020-06-04_11_30/fgtutil -I/home/sks/lxy/fos2_2020-06-04_11_30/include -I/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include -I/home/sks/lxy/fos2_2020-06-04_11_30/migbase/openssl/include -I/home/sks/lxy/fos2_2020-06-04_11_30/migbase/openssl/fn_ext -I/home/sks/lxy/fos2_2020-06-04_11_30/alohapkg/include/x86_64 -I/home/sks/lxy/fos2_2020-06-04_11_30/alohapkg/include -DCONFIG_X86_64 -DHAVE_ZEBOS -DHAVE_ZEBOS_DROUTER   -c -o /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/obj/debug-x86_64-fg/node_info.o /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:30:8: error: redefinition of 'struct in6_addr'
 struct in6_addr {
        ^~~~~~~~
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:211:8: note: originally defined here
 struct in6_addr
        ^~~~~~~~
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:36:0: error: "s6_addr" redefined [-Werror]
 #define s6_addr   in6_u.u6_addr8
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:219:0: note: this is the location of the previous definition
 #define s6_addr   __in6_u.__u6_addr8
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:37:0: error: "s6_addr16" redefined [-Werror]
 #define s6_addr16  in6_u.u6_addr16
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:221:0: note: this is the location of the previous definition
 # define s6_addr16  __in6_u.__u6_addr16
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:38:0: error: "s6_addr32" redefined [-Werror]
 #define s6_addr32  in6_u.u6_addr32
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:222:0: note: this is the location of the previous definition
 # define s6_addr32  __in6_u.__u6_addr32
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:58:8: error: redefinition of 'struct sockaddr_in6'
 struct sockaddr_in6 {
        ^~~~~~~~~~~~
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:12:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/sys/socket.h:79:17: note: originally defined here
 typedef union { __SOCKADDR_ALLTYPES
                 ^
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:66:8: error: redefinition of 'struct ipv6_mreq'
 struct ipv6_mreq {
        ^~~~~~~~~
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:288:8: note: originally defined here
 struct ipv6_mreq
        ^~~~~~~~~
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:134:0: error: "IPPROTO_HOPOPTS" redefined [-Werror]
 #define IPPROTO_HOPOPTS  0 /* IPv6 hop-by-hop options */
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:102:0: note: this is the location of the previous definition
 #define IPPROTO_HOPOPTS  IPPROTO_HOPOPTS
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:135:0: error: "IPPROTO_ROUTING" redefined [-Werror]
 #define IPPROTO_ROUTING  43 /* IPv6 routing header  */
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:104:0: note: this is the location of the previous definition
 #define IPPROTO_ROUTING  IPPROTO_ROUTING
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:136:0: error: "IPPROTO_FRAGMENT" redefined [-Werror]
 #define IPPROTO_FRAGMENT 44 /* IPv6 fragmentation header */
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:106:0: note: this is the location of the previous definition
 #define IPPROTO_FRAGMENT IPPROTO_FRAGMENT
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:137:0: error: "IPPROTO_ICMPV6" redefined [-Werror]
 #define IPPROTO_ICMPV6  58 /* ICMPv6   */
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:108:0: note: this is the location of the previous definition
 #define IPPROTO_ICMPV6  IPPROTO_ICMPV6
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:138:0: error: "IPPROTO_NONE" redefined [-Werror]
 #define IPPROTO_NONE  59 /* IPv6 no next header  */
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:110:0: note: this is the location of the previous definition
 #define IPPROTO_NONE  IPPROTO_NONE
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:139:0: error: "IPPROTO_DSTOPTS" redefined [-Werror]
 #define IPPROTO_DSTOPTS  60 /* IPv6 destination options */
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:112:0: note: this is the location of the previous definition
 #define IPPROTO_DSTOPTS  IPPROTO_DSTOPTS
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:140:0: error: "IPPROTO_MH" redefined [-Werror]
 #define IPPROTO_MH  135 /* IPv6 mobility header  */
 
In file included from /fortidev/usr/include/netdb.h:27:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/netinet/in.h:114:0: note: this is the location of the previous definition
 #define IPPROTO_MH  IPPROTO_MH
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:171:0: error: "IPV6_ADD_MEMBERSHIP" redefined [-Werror]
 #define IPV6_ADD_MEMBERSHIP 20
 
In file included from /fortidev/usr/include/netinet/in.h:37:0,
                 from /fortidev/usr/include/netdb.h:27,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/bits/in.h:234:0: note: this is the location of the previous definition
 # define IPV6_ADD_MEMBERSHIP IPV6_JOIN_GROUP
 
In file included from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:57:0,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/home/sks/lxy/fos2_2020-06-04_11_30/kernel/include/linux/in6.h:172:0: error: "IPV6_DROP_MEMBERSHIP" redefined [-Werror]
 #define IPV6_DROP_MEMBERSHIP 21
 
In file included from /fortidev/usr/include/netinet/in.h:37:0,
                 from /fortidev/usr/include/netdb.h:27,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/ping.h:23,
                 from /home/sks/lxy/fos2_2020-06-04_11_30/daemon/iputils/ping/node_info.c:37:
/fortidev/usr/include/bits/in.h:235:0: note: this is the location of the previous definition
 # define IPV6_DROP_MEMBERSHIP IPV6_LEAVE_GROUP
```

**Comment by pevik** [comment]
The support for these old systems was removed in 43afde2. Assuming you're use glibc or uclibc. You have 2 options: revert it for your build (add back `in6_flowlabel.h` and use it + maintain this patch) or patch your glibc with the fix [c9bd40daaee18cf1d9824e4a7ebaebe321e0a5a8](https://sourceware.org/git/?p=glibc.git;a=commit;h=c9bd40daaee18cf1d9824e4a7ebaebe321e0a5a8) (adapt `/usr/include/bits/in.h`).
See also https://sourceware.org/glibc/wiki/Synchronizing_Headers
